cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,387 @@
+		return nil
+	}
+
+	reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(txn.passphrase, lcm)
+	if err != nil {
+		return errors.Wrapf(err,
+			"failed to open transaction reader for ledger %d",
+			lcm.LedgerSequence())
+	}
+
+	transactions := make(map[xdr.Hash]ingest.LedgerTransaction, txCount)
+	for i := 0; i < txCount; i++ {
+		tx, err := reader.Read()
+		if err != nil {
+			return errors.Wrapf(err, "failed reading tx %d", i)
+		}
+
+		// For fee-bump transactions, we store lookup entries for both the outer
+		// and inner hashes.
+		if tx.Envelope.IsFeeBump() {
+			transactions[tx.Result.InnerHash()] = tx
+		}
+		transactions[tx.Result.TransactionHash] = tx
+	}
+
+	mid := time.Now()
+	L.WithField("passphrase", txn.passphrase).
+		Debugf("Ingesting %d transaction lookups from ledger", len(transactions))
+
+	query := sq.Insert(transactionTableName).
+		Columns("hash", "ledger_sequence", "application_order", "is_soroban")
+	for hash, tx := range transactions {
+		hexHash := hex.EncodeToString(hash[:])
+		is_soroban := (tx.UnsafeMeta.V == 3 && tx.UnsafeMeta.V3.SorobanMeta != nil)
+		query = query.Values(hexHash, lcm.LedgerSequence(), tx.Index, is_soroban)
+	}
+	_, err = query.RunWith(txn.stmtCache).Exec()
+
+	L.WithError(err).
+		WithField("duration", time.Since(start)).
+		WithField("sql_duration", time.Since(mid)).
+		Infof("Ingested %d transaction lookups", len(transactions))
+
+	if txn.ingestMetric != nil {
+		txn.ingestMetric.Observe(time.Since(start).Seconds())
+		txn.insertMetric.Observe(time.Since(mid).Seconds())
+		txn.countMetric.Observe(float64(txCount))
+	}
+
+	return err
+}
+
+func (txn *transactionHandler) RegisterMetrics(ingest, insert, count prometheus.Observer) {
+	txn.ingestMetric = ingest
+	txn.insertMetric = insert
+	txn.countMetric = count
+}
+
+// trimTransactions removes all transactions which fall outside the ledger retention window.
+func (txn *transactionHandler) trimTransactions(latestLedgerSeq uint32, retentionWindow uint32) error {
+	if latestLedgerSeq+1 <= retentionWindow {
+		return nil
+	}
+
+	cutoff := latestLedgerSeq + 1 - retentionWindow
+	_, err := sq.StatementBuilder.
+		RunWith(txn.stmtCache).
+		Delete(transactionTableName).
+		Where(sq.Lt{"ledger_sequence": cutoff}).
+		Exec()
+	return err
+}
+
+// NewTx creates a read-only SQL transaction on a cloned database session. You
+// MUST call `.Done()` on the resulting reader if there are no errors.
+func (txn *transactionHandler) NewTx(ctx context.Context) (TransactionReaderTx, error) {
+	sesh := txn.db.Clone()
+	if err := sesh.BeginTx(ctx, &sql.TxOptions{ReadOnly: true}); err != nil {
+		return nil, err
+	}
+
+	return &transactionReaderTx{ctx: ctx, db: sesh, passphrase: txn.passphrase}, nil
+}
+
+// GetLedgerRange will fetch from the database in an isolated tx.
+func (txn *transactionHandler) GetLedgerRange(ctx context.Context) ledgerbucketwindow.LedgerRange {
+	reader, err := txn.NewTx(ctx)
+	defer reader.Done()
+	if err != nil {
```

Comment:
the error should be returned so the caller can be able to interpret it

Reply:
good call, done :+1: